lib/srv/app/cloud.go

```
@@ -126,79 +128,182 @@ func (c *cloud) GetAWSSigninURL(req AWSSigninRequest) (*AWSSigninResponse, error
+		// using an OIDC token.
+		// https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/#WebIdentityRoleProvider
+		//
+		// IAM roles for EKS service accounts are also granted through the OIDC tokens.
+		// https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/
+		stscreds.WebIdentityProviderName,
+
+		// ssocreds.Provider is an AWS credential provider that retrieves
+		// temporary AWS credentials by exchanging an SSO login token.
+		// https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/ssocreds/#Provider
+		ssocreds.ProviderName:
+		return true, nil
+	}
 
-	return &AWSSigninResponse{
-		SigninURL: signinURL.String(),
-	}, nil
+	// For other providers, make an assumption that a session token is only
+	// required for temporary security credentials retrieved via STS, otherwise
+	// it is an empty string.
+	// https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#NewStaticCredentials
+	return credentials.SessionToken != "", nil
+}
+
+// getFederationDuration calculates the duration of the federated session.
+func (c *cloud) getFederationDuration(req *AWSSigninRequest, temporarySession bool) time.Duration {
+	// Max AWS federation session duration is 12 hours. The federation endpoint
+	// will error out if we request more.
+	//
+	// For AWS sessions using temporary credentials, the maximum is one hour,
+	// which is the maximum duration you can set when role chaining.
+	// https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_terms-and-concepts.html
+	maxDuration := maxSessionDuration
+	if temporarySession {
+		maxDuration = maxTemporarySessionDuration
+	}
+
+	duration := req.Identity.Expires.Sub(c.cfg.Clock.Now())
+	if duration > maxDuration {
+		duration = maxDuration
+	}
+
+	// Both the federation endpoint and the AssumeRole call will error out if
+	// we request less than 15 minutes.
+	if duration < minimumSessionDuration {
+		duration = minimumSessionDuration
+	}
+	return duration
```

Does this mean that if the user's certificate expires in less than 15 minutes, this will grant them a 15-minute session anyway? If I'm correct, I think we should rather return an error in this case.